Update and fix use templates (#4302)
influxdata/docs-v2: content/influxdb/v2.3/influxdb-templates/use.md

```sh
https://raw.githubusercontent.com/influxdata/community-templates/master/docker/docker.yml
```

{{% /note %}}

<a class="btn" href="https://github.com/influxdata/community-templates/" target="\_blank">View InfluxDB Community Templates</a>

## Use templates

Use the `influx` command line interface (CLI) and the InfluxDB `/api/v2/templates` API to summarize, validate, and apply templates.

- [View a template summary](#view-a-template-summary)
- [Validate a template](#validate-a-template)
- [Apply templates](#apply-templates)

### View a template summary

With the `influx` CLI, you can generate a summary
of template resources.
The summary contains resource metadata in table format that you can review before
applying the template to your InfluxDB instance.
You can generate a summary for a template located at a file path or at a URL.

{{% warn %}}
You can't generate a template summary with the InfluxDB `/api/v2` API.
{{% /warn %}}

#### View a template summary with the CLI

To view a template summary,
use the [`influx template` command](/influxdb/v2.3/reference/cli/influx/template/).

```sh
# Syntax

> influx template -f <FILE_PATH>

Author: ```suggestion
influx template --file <FILE_PATH>
influx template -f <FILE_PATH>
```

> influx template -u <FILE_URL>

Author: ```suggestion
influx template --template-url <FILE_URL>
influx template -u <FILE_URL>
```

```

> `-f` and `-u` aren't repeatable; you can only generate a summary for one template at a time.

Author: Thanks, I considered including both.

> `-f` and `-u` aren't repeatable; you can only generate a summary for one template at a time.

Author: ```suggestion
`--file` and `--template-url` aren't repeatable; you can only generate a summary for one template at a time.
```

The following code samples show how to generate a template summary:

- Summarize a template located on the local file system:

   ```sh
   influx template -f /PATH/TEMPLATE_FILE.yml
   ```

- Summarize a template located at a remote URL:

  ```sh
  influx template -u https://raw.githubusercontent.com/influxdata/community-templates/master/linux_system/linux_system.yml
  ```

The output is a series of tables that contain template resource metadata.

## Validate a template

Use the CLI or API to validate and troubleshoot templates before you apply them
to your InfluxDB instance.

### Validate a template with the CLI

Use the [`influx template validate` command](/influxdb/v2.3/reference/cli/influx/template/validate/) to validate a template from your local file system or a URL.

```sh
# Syntax
influx template validate -f <FILE_PATH>
influx template validate -u <FILE_URL>
```

To validate a template stored on your file system, enter the following command into the terminal:

```sh
influx template validate -f /PATH/TO/TEMPLATE.yml
```

Replace **`/PATH/TO/TEMPLATE.yml`** with the file path for your template.

> To validate a template located at a URL, pass the `u` option with the URL.

Author: I've typically used "flag" to mean boolean-only options.
That interpretation is fairly common in Unix/Linux, I think.
And then other "options" or "parameters" can take values (or "arguments").

InfluxDB validates and then installs the template to create and update resources in the organization.
If you apply a template without providing a **stack ID**, InfluxDB initializes a new stack with all new resources.

To learn how to use InfluxDB stacks, see [InfluxDB stacks](/influxdb/v2.3/influxdb-templates/stacks/).

After applying the template, InfluxDB responds with the following:

- the stack ID for the stack that was created or updated
- the diff of added and removed resources

{{% note %}}
When applying templates, you must pass a token that has `write` permission for the
resource types configured in the template.
{{% /note %}}

Use the `influx` CLI or `/api/v2` API to apply templates to an InfluxDB instance.

- [Apply templates with the CLI](#apply-templates-with-the-cli)
- [Apply templates with the API](#apply-templates-with-the-api)

## Apply templates with the CLI

Use the [`influx apply` command](/influxdb/v2.3/reference/cli/influx/apply/) to install templates to your InfluxDB organization.
You can apply templates stored in your local file system or from URLs.

The following code sample uses `influx apply` with an

> [all-access token]({{< latest "influxdb" >}}/security/tokens/) to

Reviewer: We should use the `latest` shortcode to reference the same product/version. This will lead to inadvertent version switches when a new version gets released. 

```suggestion
[all-access token](/influxdb/v2.3/security/tokens/) to
```

|                                                                                 TOTAL              |      1      |
+--------------------+------------------+---------------+-----------+--------------------------------+-------------+

Stack ID: 09bd87cd33be3000
```

- [Apply a template from the file system](#apply-a-template-from-the-file-system)
- [Apply all templates in a directory](#apply-all-templates-in-a-directory)
- [Apply a template from a URL](#apply-a-template-from-a-url)
- [Apply templates from both files and URLs](#apply-templates-from-both-files-and-urls)
- [Apply templates from files and URLs](#apply-templates-from-files-and-urls)
- [Define environment references](#define-environment-references)
- [Include a secret when installing a template](#include-a-secret-when-installing-a-template)

{{% note %}}
#### Apply templates to an existing stack
To apply a template to an existing stack, include the stack ID when applying the template.
Any time you apply a template without a stack ID, InfluxDB initializes a new stack
and all new resources.
For more information, see [InfluxDB stacks](/influxdb/v2.3/influxdb-templates/stacks/).

> {{% /note %}}

Reviewer: Why remove this note? This is important when applying templates and you want to update an existing stack with an additional tempalte.

Author: > Why remove this note? This is important when applying templates and you want to update an existing stack with an additional tempalte.

Removed the note and moved it to https://github.com/influxdata/docs-v2/pull/4302/files#diff-b9fc01b5b06eb5bff2b5705f8c6d908b75f0ddbff4dad75f52bfba9957d6282bR242

Reviewer: Oh gotcha. I just read to quickly and missed over the new section. Thanks.

```

To use one command to apply multiple templates from URLs, pass the `-u` option
and the URL for each template--for example:

```sh
# Apply multiple templates from URLs
influx apply -o example-org \
  -u https://example.com/templates/template1.yml \
  -u https://example.com/templates/template2.yml
```

### Apply multiple templates from files and URLs

To use a single command to apply templates from multiple file paths and URLs, pass each template with the appropriate `-f` or `-u` option.

```sh
# Syntax
influx apply -o <INFLUX_ORG> -u <FILE_URL> -f <FILE_PATH>
```

The following code sample shows a single command to apply multiple templates from URLs and file paths:

```sh
influx apply -o INFLUX_ORG \
  -u https://example.com/templates/template1.yml \
  -u https://example.com/templates/template2.yml \
  -f ~/templates/custom-template.yml \
  -f ~/templates/iot/home/ \
  --recurse
```

The `--recurse` flag applies any templates from the `~/templates/iot/home/`
directory and subdirectories.

## Apply templates with the API

Use the `/api/v2/templates/apply` endpoint to install templates to your InfluxDB instance.

> [{{< api-endpoint method="POST" endpoint="/api/v2/templates/apply" >}}]({{% latest "influxdb" %}}/api/#operation/ApplyTemplate)

Reviewer: ```suggestion
[{{< api-endpoint method="POST" endpoint="/api/v2/templates/apply" >}}](/influxdb/v2.3/api/#operation/ApplyTemplate)
```

              "color": "#326BBA",
              "name": "inputs.cpu"
            }
          }]
        },
        { "contents": [{
            "apiVersion": "influxdata.com/v2alpha1",
            "kind": "Bucket",
            "metadata": {
              "name": {
                "envRef": {
                  "key": "docker-bucket"
                }
              }
            },
            "spec": {
              "name": {
                "envRef": {
                  "key": "docker-spec-1"
                }
              },
              "retentionRules": [
                {
                  "everySeconds": 604800,
                  "type": "expire"
                }
              ]
            }
          }]
        }
      ]
    }
EOF
```

{{% /truncate %}}

For more API schema detail, see the [`/api/v2/templates/apply` reference documentation](/influxdb/v2.3/api/#operation/ApplyTemplate).

## Pass secrets when installing a template

Some templates use [secrets](/influxdb/v2.3/security/secrets/) in queries.

- [Pass secrets with the CLI](#pass-secrets-with-the-cli)
- [Pass secrets with the API](#pass-secrets-with-the-api)

### Example template with a secret

> For example, the following `wins` task (from the **fortnite** community template) contains a `query`

Author: ```suggestion
The following sample `wins` task (from the **fortnite** community template) contains a `query`
```

    "labelMappings": [],
    "missingEnvRefs": null,
    "missingSecrets": null,
    "summaryTask": [],
    "telegrafConfigs": [],
    "variables": []
  },
  "errors": [
    {
      "kind": "Foo",
      "fields": [
        "root",
        "kind"
      ],
      "idxs": [
        0,
        null
      ],
      "reason": "unsupported kind provided"
    }
  ],
  "code": "unprocessable entity",
  "message": "unprocessable entity"
}
```

{{% /truncate %}}

Given that the template contains an invalid InfluxDB resource type (`"Foo"`),
the template fails validation and the response contains the following:

- empty resource diffs
- list of errors
- error code and message

## Apply templates to an InfluxDB instance

When you apply a template, you specify the organization that owns the template resources.
InfluxDB validates and then installs the template to create and update resources in the organization.
If you apply a template without providing a **stack ID**, InfluxDB initializes a new stack with all new resources.

To learn how to use InfluxDB stacks, see [InfluxDB stacks](/influxdb/v2.3/influxdb-templates/stacks/).

After applying the template, InfluxDB responds with the following:

> - stack ID for the stack that was created or updated
> - diff of added and removed resources

Author: ```suggestion
- Stack ID for the stack that was created or updated
- A diff of added and removed resources
```

### Define environment references with the CLI

The `influx apply` command prompts you to provide a value for each environment
reference in the template.
You can also provide values for environment references by passing an `--env-ref`
option with a key-value pair comprised of the environment reference key and the
value to replace it--for example:

```sh
influx apply -o INFLUX_ORG_ID -f /path/to/template.yml \
  --env-ref=bucket-name-1=myBucket
  --env-ref=label-name-1=Label1 \
  --env-ref=label-name-2=Label2
```

### Define environment references with the API

To provide environment reference values when you apply a template with the
`/api/v2/templates/apply` API endpoint, pass _`envRefs`_ with key-value pairs
in the request body.

The following code sample shows a request that defines environment references (`envRef`) in template objects,
and passes environment reference values (`envRefs`):

{{% truncate %}}

```sh
  curl -v "http://localhost:8086/api/v2/templates/apply" \
    --header "Authorization: Token INFLUX_API_TOKEN" \
    --data @- << EOF
    { "orgID": "INFLUX_ORG_ID",
      "dryRun": true,
      "envRefs": {

>         "linux-cpu-label": "MY-CPU-LABEL",
>         "docker-bucket": "MY-DOCKER-BUCKET",
>         "docker-spec-1": "MY-DOCKER-SPEC"

Reviewer 2: ```suggestion
        "linux-cpu-label": "MY_CPU_LABEL",
        "docker-bucket": "MY_DOCKER_BUCKET",
        "docker-spec-1": "MY_DOCKER_SPEC"
```

```

_To add a secret after applying a template, see [Add secrets](/influxdb/v2.3/security/secrets/manage-secrets/add/)._

### Pass secrets with the API

To define secret values when installing a template with the InfluxDB `/api/v2/templates/apply` endpoint, pass the `secrets` parameter in the request body with an object that contains key-value pairs.

The following code sample shows how to pass `SLACK_WEBHOOK` in `secrets` when applying the [**fortnite** template](#example-template-with-a-secret):

```sh
  curl "http://localhost:8086/api/v2/templates/apply" \
    --header "Authorization: Token INFLUX_API_TOKEN" \
    --data @- << EOF | jq .
      {
        "dryRun": false,
        "orgID": "INFLUX_ORG_ID",
        "secrets": {
          "SLACK_WEBHOOK": "YOUR SECRET WEBHOOK URL"
        },
        "remotes": [
          {
            "url": "https://raw.githubusercontent.com/influxdata/community-templates/master/fortnite/fn-template.yml"
          }
        ]
      }
EOF
```

The secret value is not exposed in the InfluxDB `/api/v2/templates/apply` API response.

> For more detail about the API schema, see the [`/api/v2/templates/apply` reference documentation]({{% latest "influxdb" %}}/api/#operation/ApplyTemplate).

Reviewer: ```suggestion
For more detail about the API schema, see the [`/api/v2/templates/apply` reference documentation](/influxdb/v2.3/api/#operation/ApplyTemplate).
```

```

_To add a secret after applying a template, see [Add secrets](/influxdb/v2.3/security/secrets/manage-secrets/add/)._

### Pass secrets with the API

To define secret values when installing a template with the InfluxDB `/api/v2/templates/apply` endpoint, pass the `secrets` parameter in the request body with an object that contains key-value pairs.

The following code sample shows how to pass `SLACK_WEBHOOK` in `secrets` when applying the [**fortnite** template](#example-template-with-a-secret):

```sh
  curl "http://localhost:8086/api/v2/templates/apply" \
    --header "Authorization: Token INFLUX_API_TOKEN" \
    --data @- << EOF | jq .
      {
        "dryRun": false,
        "orgID": "INFLUX_ORG_ID",
        "secrets": {

>           "SLACK_WEBHOOK": "YOUR SECRET WEBHOOK URL"

Reviewer 2: ```suggestion
          "SLACK_WEBHOOK": "YOUR_SECRET_WEBHOOK_URL"
```
